From: Hannes Reinecke <hare@suse.de>
Date: Thu, 24 May 2018 16:18:17 +0200
Subject: [PATCH] nvme-fabrics: allow duplicate connections to the discovery
 controller
References: bsc#1098527
Git-commit: 181303d03525ea52d2d002fb8ee04e769aaa4ce4
Patch-mainline: v4.18-rc1

The whole point of the discovery controller is that it can accept
multiple connections.

[hare: ported to SLE12 SP3]

Signed-off-by: Hannes Reinecke <hare@suse.com>
Reviewed-by: James Smart <james.smart@broadcom.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 drivers/nvme/host/fabrics.c | 1 +
 drivers/nvme/host/fc.c      | 1 +
 2 files changed, 2 insertions(+)

diff --git a/drivers/nvme/host/fabrics.c b/drivers/nvme/host/fabrics.c
index 27ad2b4d1d32..f9adf660c9f0 100644
--- a/drivers/nvme/host/fabrics.c
+++ b/drivers/nvme/host/fabrics.c
@@ -860,6 +860,7 @@ static int nvmf_parse_options(struct nvmf_ctrl_options *opts,
 	if (opts->discovery_nqn) {
 		opts->kato = 0;
 		opts->nr_io_queues = 0;
+		opts->duplicate_connect = true;
 	}
 	if (ctrl_loss_tmo < 0)
 		opts->max_reconnects = -1;
diff --git a/drivers/nvme/host/fc.c b/drivers/nvme/host/fc.c
index 26a10fff6029..ee6245c33c45 100644
--- a/drivers/nvme/host/fc.c
+++ b/drivers/nvme/host/fc.c
@@ -3130,6 +3130,7 @@ nvme_fc_init_ctrl(struct device *dev, struct nvmf_ctrl_options *opts,
 	}
 
 	if (!opts->duplicate_connect &&
+	    !opts->discovery_nqn &&
 	    nvme_fc_existing_controller(rport, opts)) {
 		ret = -EALREADY;
 		goto out_fail;
-- 
2.12.3

